source/helpers/conditionParser.ts

```diff
@@ -0,0 +1,119 @@
+import { Question, SummaryItem } from '../types/FormTypes';
+
+const evaluateSummaryList = (
+  answers: Record<string, any>,
+  items: SummaryItem[]
+): boolean => {
+  if (!Array.isArray(items)) {
+    return false;
+  }
+  let isEmpty = true;
+  items.forEach((item) => {
+    if (!isEmpty) return;
+
+    if (['text', 'number', 'date', 'checkbox'].includes(item.type)) {
+      if (answers[item.id] && answers[item.id] !== '') isEmpty = false;
+    }
+    if (['arrayText', 'arrayNumber', 'arrayDate'].includes(item.type)) {
+      if (answers[item.id] && Array.isArray(answers[item.id]) && answers[item.id].length > 0) {
+        (answers[item.id] as string[]).forEach(a => { if (a && a !== '') isEmpty = false; })
+      }
+    }
+  });
+  return !isEmpty;
+};
+
+/** Evaluates an answer value to a boolean. False if the value is empty or false, otherwise true. */
+export const evaluateAnswer = (
+  questionId: string,
+  answers: Record<string, any>,
+  questions: Question[]
+): boolean => {
+  const question = questions.find((q) => q.id === questionId);
+  if (!question) return false;
+
+  switch (question.type) {
+    case 'checkbox':
+      if (answers[questionId]) return answers[questionId];
+      return false;
+    case 'text':
+    case 'number':
+    case 'date':
+      return (answers[questionId] && answers[questionId] !== '');
+    case 'editableList':
+      return (
+        answers[questionId] &&
+        Object.keys(answers[questionId]).filter(
+          (key) => answers[questionId][key] && answers[questionId][key] !== ''
+        ).length > 0
+      );
+    case 'repeaterField':
+      return (answers[questionId] && answers[questionId].length > 0);
+    case 'summaryList':
+      return evaluateSummaryList(answers, question.items || []);
+    default:
+      return false;
+  }
+};
+
+type EvaluatedValue = boolean | '!' | '&&' | '||';
+
+// evaluation functions
+const evaluateNot = (array: EvaluatedValue[]) => {
+  const arrCopy = [...array];
+  const reversedIndex = arrCopy.reverse().findIndex(expr => expr === '!');
+  if (reversedIndex === -1) return array;
+  
+  const index = arrCopy.length - reversedIndex - 1;
+  arrCopy.reverse()[index+1] = !arrCopy[index+1];
+  arrCopy.splice(index,1);
+  return evaluateNot(arrCopy);
+};
+const evaluateAnd = (array: EvaluatedValue[]) => {
+  const arrCopy = [...array];
+  const index = arrCopy.findIndex(expr => expr === '&&');
+  if (index === -1) return array;
+  
+  arrCopy[index-1] = (array[index-1] && array[index+1]) as EvaluatedValue;
+  arrCopy.splice(index,2);
+  return evaluateAnd(arrCopy);
+}
+const evaluateOr = (array: EvaluatedValue[]): boolean[] => {
+  const arrCopy = [...array];
+  const index = arrCopy.findIndex(expr => expr === '||');
+  if (index === -1) return (array as boolean[]);
+  
+  arrCopy[index-1] = (array[index-1] || array[index+1]) as EvaluatedValue;
+  arrCopy.splice(index,2);
+  return evaluateOr(arrCopy);
+}
```

Comment:
I kind of think all sort of evaluation like this is going to be somewhat hard to understand, simply by the nature of what you're doing. Predicate functions are probably nice, but evaluating an array like this using them is also not going to be trivial to understand, especially if it's the first time seeing them. 

If one just wants the simplest possible evaluation, we could of course just use eval() after replacing all expressions with true/false and joining the array to a string again, that will work and be easy to understand. I wrote these evaluation functions because I thought it was sort of a fun little project, more than anything else. 

Also, what part is redundant?

Reply:
Redundant in all functions:
```
// Could all be broken out to a single function (findExpression)
const index = arrCopy.findIndex(expr => expr === '&&');
const index = arrCopy.findIndex(expr => expr === '||');
// Do not know the purpose for reversing the array here.
const reversedIndex = arrCopy.reverse().findIndex(expr => expr === '!');

// Same condition and return? exception in the evaluateOr
if (index === -1) return array;
if (index === -1) return (array as boolean[]);
```


Redundant code in evaluateAnd and evaluateOr:
```
arrCopy[index-1] = (array[index-1] || array[index+1]) as EvaluatedValue;
arrCopy.splice(index,2);
```

Reply:
Reversing is done at the start of the evaluateNot, and the reverse there is to go back to normal order. The purpose of reversing is that it's a valid expression to write stuff like !!value , i.e. multiple not in a row, and this is a "clever" way of allowing for that, since the last ! should be next to a boolean value, whereas the first might be next to another !. But of course in this context, it's not something you would ever really write, I was just being "clever" and allowing for an edge case. 

And yeah, the other things could be broken out into functions, sure.

Reply:
Okay so, looked at this, and broke out the evaluation to its own function. But honestly, I'm not sure this improves readability much, since you now have to understand this function as well; it sort of breaks up the flow when reading. And while it does reduce redundancy a little, it's not very much. 

And I disagree with creating a findExpression function, since what we need already exists and is called findIndex, which I already use. Wrapping a oneliner that's using a standard library function in its own tiny function feels like just creating more redundancy, not reducing it.

Reply:
"...it sort of breaks up the flow when reading. And while it does reduce redundancy a little, it's not very much."
 You have a point here, from my point of view these changes are not necessary. Just a observation, that's why i left is as a comment and not a as a change request (y) 
 
 "And I disagree with creating a findExpression function, since what we need already exists and is called findIndex, which I already use. "
My suggestion here was vague, where i was going here is not to only replace the findIndex function but rather have a function that wraps these two lines:
```
const index = arrCopy.findIndex(expr => expr === '&&');
if (index === -1) return array;

function findOperator(condition, operator) {
  const index = condition.findIndex(conditionValue => conditionValue === operator);
  
  // This part is just an example, you could ofcourse return something more suitable.
  if (index === -1) return [false, index];
  return [true, index]
}
```